### ISSUE TITLE
Sparse matrix multiply orders of magnitude slower than PyTorch

### ISSUE THREAD

**Issue by cjnolet** [cupy/cupy#2664]
I have been doing some profiling comparing cupy's sparse matrix `dot()` with PyTorch's `mm()` and I'm getting some very surprising results. I noticed that while cupy is using cusparse, PyTorch appears to be using only thrust/cub to do the multiply. 

Strangely, I'm finding that it's the cusparse `csrgemm_kernel` from cupy that is dominating the runtime of the multiply, taking 40ms. The end-to-end matrix multiply is taking upwards of 70ms for cupy. The PyTorch implementation is taking a little over 2ms to run the multiply end-to-end. 

I'm attaching the [nvvp file](https://github.com/cupy/cupy/files/3870573/cupy_vs_pytorch.nvvp.zip) with my profiling results. I'm pretty surprised. The algorithm being profiled is a simple Multinomial Naive Bayes using the 20-newsgroups dataset from scikit-learn. I trained both algorithms twice in order to eliminate jit and cuda context creation from the comparison.

```
>>> cupy.show_config()
CuPy Version          : 6.5.0
CUDA Root             : /usr/local/cuda
CUDA Build Version    : 10000
CUDA Driver Version   : 10000
CUDA Runtime Version  : 10000
cuDNN Build Version   : 7600
cuDNN Version         : 7600
NCCL Build Version    : 2406
NCCL Runtime Version  : 2406
```

**Comment by niboshi** [comment]
Related: #2664

**Comment by kmaehashi** [comment]
Could you try the proposed patch? https://github.com/cupy/cupy/pull/2698

**Comment by grlee77** [comment]
> Could you try the proposed patch? #2698

I only implemented matrix-vector multipy (spmv) in #2698 and did not try spmm. (Also the gains or not an order of magnitude as mentioned here).  Regardless of how small the sparse matrix is, it seems that even the CUB-based approach still takes close to 20 ms. It could be useful to determine what the source of that overhead is.

@cjnolet.  How many nonzero entries are in the dataset you are testing with?

**Comment by cjnolet** [comment]
@grlee77, I'm using the 20-newgroups with all the labels from scikit-learn, which looks to have an nnz of 1787565.

```python
from sklearn.metrics import accuracy_score
from sklearn.datasets import fetch_20newsgroups
from sklearn.feature_extraction.text import CountVectorizer
from sklearn.preprocessing import LabelBinarizer

import cupy.prof
import time
import math
import torch
import numpy as np

```

Here's the PyTorch Naive Bayes implementation that I've been using for comparison:
```python
def scipy_to_torch(sp):
    coo = sp.tocoo()
    values = coo.data
    indices = np.vstack((coo.row, coo.col))

    i = torch.cuda.LongTensor(indices)
    v = torch.cuda.FloatTensor(values)

    return torch.cuda.sparse.FloatTensor(i, v, torch.Size(coo.shape))

class PyTorchBayes(object):

    def __init__(self, l, alpha=1.0, fit_prior=True, class_prior=None):
        self.alpha = alpha
        self.l = l
        self.fit_prior = fit_prior
        self.class_prior = class_prior

        self.n_features_ = None

    @cp.prof.TimeRangeDecorator(message="pytorch_fit()", color_id=1)
    def fit(self, X, y, _partial=False, _classes=None):

        self.n_features_ = X.shape[1]
        self._init_counters(y.shape[1], X.shape[1])

        self.classes_ = self.l.classes_
        self.n_classes_ = self.l.classes_.shape[0]

        self._count(X, y)
        self._update_feature_log_prob(self.alpha)
        self._update_class_log_prior(class_prior=self.class_prior)

        return self

    def predict(self, X):
        jll = self._joint_log_likelihood(X)

        _, indices = torch.max(jll, 1)
        return indices

    def _init_counters(self, n_effective_classes, n_features):
        self.class_count_ = torch.zeros(n_effective_classes).cuda()
        self.feature_count_ = torch.zeros(n_effective_classes, n_features).cuda()

    def _count(self, X, Y):

        with cp.prof.time_range(message="pytorch_matrix_multiply", color_id=7):
            feature_count_ = torch.sparse.mm(X.t(), Y).t()

        print(str(feature_count_))

        self.feature_count_ += feature_count_
        self.class_count_ += Y.sum(axis=0)

    def _update_class_log_prior(self, class_prior=None):

        if class_prior is not None:
            self.class_log_prior_ = torch.log(class_prior)

        elif self.fit_prior:
            log_class_count = torch.log(self.class_count_)

        self.class_log_prior_ = torch.full((self.n_classes_, 1),
                                           -math.log(self.n_classes_)).cuda()

    def _update_feature_log_prob(self, alpha):
        """ apply add-lambda smoothing to raw counts and recompute log probabilities"""
        smoothed_fc = self.feature_count_ + alpha
        smoothed_cc = smoothed_fc.sum(axis=1).reshape(-1, 1)
        self.feature_log_prob_ = (torch.log(smoothed_fc) - torch.log(smoothed_cc))

    def _joint_log_likelihood(self, X):
        """ Calculate the posterior log probability of the samples X """
        ret = torch.sparse.mm(X, self.feature_log_prob_.T)
        ret += self.class_log_prior_.T
        return ret

def load_corpus_cpu():

    twenty_train = fetch_20newsgroups(subset='train',
                                      shuffle=True, random_state=42)

    count_vect = CountVectorizer()
    X = count_vect.fit_transform(twenty_train.data)
    Y = twenty_train.target

    return X, Y
```

And here's the small script I'm running to test PyTorch:

```python

X, y = load_corpus_cpu()
l = LabelBinarizer()
Y = torch.cuda.FloatTensor(l.fit_transform(y)).cuda()

a = scipy_to_torch(X)

# Run once to "prime" any initialization
m = PyTorchBayes(l)
m.fit(a, Y)

with cupy.prof.time_range(message="fit()"):
  start = time.time()
  m = PyTorchBayes(l)
  m.fit(a, Y)
  end = time.time() - start

print("PYTORCH: %s" % str(end))

y_hat_gpu = m.predict(a)
print(str(accuracy_score(y, y_hat_gpu.cpu().numpy())))
```

I'm getting this as my output (accuracy matches scikit-learn):
```
PYTORCH: 0.008881330490112305
0.9245182959165635
```

**Comment by leofang** [comment]
`time.time()` does not include GPU activities, so looks like only kernel launch time is counted. How did you time the CuPy code?

**Comment by cjnolet** [comment]
I timed each algorithm internally using ‘cupy.prof.time_range’ (as can be seen [here](https://github.com/cupy/cupy/files/3870573/cupy_vs_pytorch.nvvp.zip) and found the numbers described in the description.

The end to end times were computed for internal cuml benchmarking, but I felt it relevant to post them here. I had run a test-case where I synchronized the streams before calling ‘start = time.time()’ but didn’t find any difference in the runtimes. 

Also, important to note- the order of magnitude difference being referred in the title is the actual runtime of the sparse multiply kernels (from the attached nvvp session) and not the end to end from ‘time.time()’.

**Comment by cjnolet** [comment]
@leofang, I missed the ‘sync=True’ in the fit() function when constructing the reproducible script. I’ve updated my previous reply accordingly.

**Comment by leofang** [comment]
Thanks, I didn't pay attention to the origin post and jumped into the middle of the thread. Could you share your cupy script so we can test locally ourselves? It'd offer a different view from an nvvp session.

**Comment by smarkesini** [comment]
sorry that I opened a different issue for SpMV, but in #2790 I attached a modification to cupyx/scipy/sparse/coo.py, that improved the timing for SpMV with 10M nnz  from 0.06 to 0.00024 seconds. I could be related.
